cartographer_ros/cartographer_ros/assets_writer.cc

```diff
@@ -189,73 +191,100 @@ void AssetsWriter::Run(const std::string& configuration_directory,
       if (trajectory_proto.node_size() == 0) {
         continue;
       }
-      tf2_ros::Buffer tf_buffer;
+      std::shared_ptr<tf2_ros::Buffer> tf_buffer = std::make_shared<tf2_ros::Buffer>(clock);
+      tf_buffer->setUsingDedicatedThread(true);
       if (!urdf_filename.empty()) {
-        ReadStaticTransformsFromUrdf(urdf_filename, &tf_buffer);
+        ReadStaticTransformsFromUrdf(urdf_filename, tf_buffer);
       }
-
       const carto::transform::TransformInterpolationBuffer
           transform_interpolation_buffer(trajectory_proto);
-      rosbag::Bag bag;
-      bag.open(bag_filename, rosbag::bagmode::Read);
-      rosbag::View view(bag);
-      const ::ros::Time begin_time = view.getBeginTime();
-      const double duration_in_seconds =
-          (view.getEndTime() - begin_time).toSec();
+      rosbag2_cpp::Reader bag_reader(
+        std::make_unique<rosbag2_cpp::readers::SequentialReader>()
+      );
+      rosbag2_storage::StorageOptions storage_options {
+          bag_filename,
+          "sqlite3",
+      };
+      rosbag2_cpp::ConverterOptions converter_options {
+          "cdr",
+          "cdr"
+      };
+      bag_reader.open(storage_options, converter_options);
+      rosbag2_storage::BagMetadata meta_data = bag_reader.get_metadata();
+      auto begin_time = meta_data.starting_time;
+
+      std::unordered_map<std::string, std::string> topic_name_to_type;
```

Comment:
Make sure to `#include <unordered_map>` at the top for this.

Reply:
Done